gemd/util/impl.py

```diff
@@ -98,34 +109,118 @@ def _substitute(thing: Any,
         visited = {}
     if thing.__hash__ is not None and thing in visited:
         return visited[thing]
+
     if applies(thing):
         replacement = sub(thing)
         if thing.__hash__ is not None:
             visited[thing] = replacement
-        new = _substitute(replacement, sub, applies, visited)
-    elif isinstance_list(thing):
-        new = [_substitute(x, sub, applies, visited) for x in thing]
-    elif isinstance_tuple(thing):
-        new = tuple(_substitute(x, sub, applies, visited) for x in thing)
-    elif isinstance_dict(thing):
+    else:
+        replacement = thing
+
+    if _cached_isinstance(replacement, list):
+        new = [_substitute(x, sub, applies, visited) for x in replacement]
+    elif _cached_isinstance(replacement, tuple):
+        new = tuple(_substitute(x, sub, applies, visited) for x in replacement)
+    elif _cached_isinstance(replacement, dict):
         new = {_substitute(k, sub, applies, visited): _substitute(v, sub, applies, visited)
-               for k, v in thing.items()}
-    elif isinstance_dict_serializable(thing):
+               for k, v in replacement.items()}
+    elif _cached_isinstance(replacement, DictSerializable):
         new_attrs = {_substitute(k, sub, applies, visited): _substitute(v, sub, applies, visited)
-                     for k, v in thing.as_dict().items()}
-        new = thing.build(new_attrs)
+                     for k, v in replacement.as_dict().items()}
+        new = replacement.build(new_attrs)
     else:
-        new = thing
+        new = replacement
 
     if thing.__hash__ is not None:
         visited[thing] = new
-    if new.__hash__ is not None:
-        visited[new] = new
 
-    # assert type(thing) == type(new), "{} is not {}".format(type(thing), type(new))
     return new
 
 
+def _substitute_inplace(thing: Any,
```

Comment:
Do we have support for deep-copying GEMD objects?

My naive way to implement this would be to have `_substitute_inplace`, and then have `_substitute` call a copy of the passed object, and then in-place substitute on the copy.

In terms of code complexity, this reduces the problem to a single implementation. But that may defeat the performance aspects in some way if copying is too expensive.

Reply:
I agree with the sentiment about having a large number of methods that perform similar tasks, but the deep copy methods go through JSON serialization, which is a remarkably slow task.  There's a lot of space for housecleaning in this domain across gemd-python & citrine-python, but I don't think that refactor is quite justified yet.